### PR TITLE
Numbers compared to enums should perform Varchar comparison

### LINKED ISSUE #8561: R: Comparing factors with numbers leads to unexpected results
### What happens?

A factor (enum) never seems to be equal to a number in the relational API. This should either work or give an error.

### To Reproduce

``` r
con <- DBI::dbConnect(duckdb::duckdb())
experimental <- FALSE
invisible(DBI::dbExecute(con, "CREATE MACRO \"==\"(a, b) AS a = b"))
df1 <- data.frame(a = c("a", "b"), b = factor(c("1", "2")))

rel1 <- duckdb:::rel_from_df(con, df1, experimental = experimental)

# Fails:
rel2 <- duckdb:::rel_filter(
  rel1,
  list(
    duckdb:::expr_function(
      "==",
      list(duckdb:::expr_reference("b"), duckdb:::expr_constant(1))
    )
  )
)
rel2
#> DuckDB Relation: 
#> ---------------------
#> --- Relation Tree ---
#> ---------------------
#> Filter [==(b, 1.0)]
#>   r_dataframe_scan(0x1351493c8)
#> 
#> ---------------------
#> -- Result Columns  --
#> ---------------------
#> - a (VARCHAR)
#> - b (ENUM(''1'', ''2''))
duckdb:::rel_to_altrep(rel2)
#> [1] a b
#> <0 rows> (or 0-length row.names)

# Fails too:
rel3 <- duckdb:::rel_filter(
  rel1,
  list(
    duckdb:::expr_function(
      "==",
      list(duckdb:::expr_reference("b"), duckdb:::expr_constant(1L))
    )
  )
)
rel3
#> DuckDB Relation: 
#> ---------------------
#> --- Relation Tree ---
#> ---------------------
#> Filter [==(b, 1)]
#>   r_dataframe_scan(0x1351493c8)
#> 
#> ---------------------
#> -- Result Columns  --
#> ---------------------
#> - a (VARCHAR)
#> - b (ENUM(''1'', ''2''))
duckdb:::rel_to_altrep(rel3)
#> [1] a b
#> <0 rows> (or 0-length row.names)

# Works as expected:
rel4 <- duckdb:::rel_filter(
  rel1,
  list(
    duckdb:::expr_function(
      "==",
      list(duckdb:::expr_reference("b"), duckdb:::expr_constant("1"))
    )
  )
)
rel4
#> DuckDB Relation: 
#> ---------------------
#> --- Relation Tree ---
#> ---------------------
#> Filter [==(b, '1')]
#>   r_dataframe_scan(0x1351493c8)
#> 
#> ---------------------
#> -- Result Columns  --
#> ---------------------
#> - a (VARCHAR)
#> - b (ENUM(''1'', ''2''))
duckdb:::rel_to_altrep(rel4)
#>   a b
#> 1 a 1
```

<sup>Created on 2023-08-13 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### OS:

macOS aarch64

### DuckDB Version:

a8ce02cc2e740d8973d26ccdb77d0068c69c9124

### DuckDB Client:

R

### Full Name:

Kirill Müller

### Affiliation:

cynkra GmbH

### Have you tried this on the latest `master` branch?

I have tested with a master build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by Tmonster**
fixes https://github.com/duckdb/duckdb/issues/8561

Sometimes a user wishes to compare a numeric type with an enum type. Duckdb treats enums as strings, so we should be convert the enum to string, this is fast since it is a simple vector lookup for each enum. So for comparisons with Enums, we make the `MaxLogicalType` VARCHAR. this will also cast the int to VARCHAR as well. 

Both the integer and the enum are now cast to varcher and the comparison is done. Except in the case where constant integers is used. When this happens the comparison simplification rule kicks in and casts the integer to varchar and then to enum so that enum comparison/integer comparison can be used. 

Added some mini benchmarks so that we will be notified if someone adds functionality results in a regression when doing enum comparisons with casting.

### CODE REVIEW

**Comment by Tishj** [comment]
```
select 'happy' > 'sad', 'ok' > 'sad';
--┌───────────────────┬────────────────┐
--│ ('happy' > 'sad') │ ('ok' > 'sad') │
--│      boolean      │    boolean     │
--├───────────────────┼────────────────┤
--│ false             │ false          │
--└───────────────────┴────────────────┘
```

What is happening here?
If we cast these to enums then `happy` should have a higher enum code than `sad` no?

**Comment by Tmonster** [comment]
Ah, but then again, this only happens when a user really wants to compare the value of their enums. If that's the case they should explicitly cast the enum column to varcher since duckdb should always prefer integer comparisons.

File: test/sql/types/enum/test_enum_to_numbers.test [hunk 1]
```diff
@@ -0,0 +1,124 @@
+# name: test/sql/types/enum/test_enum_to_numbers.test
+# description: ENUM types used in table tests
+# group: [enum]
+
+statement ok
+create type enum1 as enum ('1', '2', '3', '4');
```
Comment (pdet) [inline test/sql/types/enum/test_enum_to_numbers.test:6]: Can we also test with other data types?

Comment (Tmonster) [reply test/sql/types/enum/test_enum_to_numbers.test:6]: Enums can only hold varchar values, do you mean compare the enum with other types?

Comment (pdet) [reply test/sql/types/enum/test_enum_to_numbers.test:6]: These are integers now, right? I imagine similar comparisons could happen with other data types, hugeints, doubles, dates, times?

**Review by pdet — CHANGES_REQUESTED** [review]
The code looks good, I would just add a test that checks the comparison of enums and other types `foreach types ints, doubles, dates, ...` Including the ones that are not supported with a `statement error` message. 
I would also add a couple more tests with comparisons that are not equality or inequality (e.g., < <= > >=)

**Review by pdet — APPROVED** [review]

**Comment by Mytherin** [comment]
Thanks! LGTM

**Pull request merged**